### PR TITLE
Disclosure Panel and Banner across Automate

### PR DESCRIPTION
**Pull request by SEAjamieD**
### :nut_and_bolt: Description: What code changed, and why?
Some of our customers have been requesting a way to add a disclosure statement to the login page, as well as a ribbon across the top of the login page and Automate in order to notify users of various message that relate to using our product on their network.  This update will allow a user to define: 

A) if they should show the disclosure text
B) Populate the disclosure with HTML content
C) if they should show the ribbon
D) Populate the ribbon with text, background color, and text color

### :chains: Related Resources
Addresses: https://github.com/chef/automate/issues/4934 and https://github.com/chef/automate/issues/4974

### :+1: Definition of Done
A user can choose to display a banner across the top of the screen in their login flow
 - They may provide their own text in the banner
 - They may choose the background color and text color

A user can choose to display disclosure text on their login page
 - They may populate the disclosure with HTML of their choice

Inside Automate
- Users may update configuration files so that they can show a customizable banner and disclosure text
- Unit tests completed that appropriately test the Warning Banner Component
- Screen shot showing what to expect when banner is present
- Screen shot showing what to expect when disclosure is present
- No visual changes to Automate UI if a user choose to not show the banner or the disclosure text.

### :athletic_shoe: How to Build and Test the Change

You'll have to rebuild 5 services for this to populate correctly. 
The related services are:

1. automate-deployment
2. automate-cli
3. automate-load-balancer
4. automate-gateway
5. automate-dex

In hab studio:
`build components/automate-deployment`
`build components/automate-cli`
`build components/automate-load-balancer`
`build components/automate-gateway`
`build components/automate-dex`
`start_all_services`

You'll also need to run `chef-automate config patch dev/patch.toml` in hab studio.  This will connect all the variables we have coming from the config go files.

If you are already logged into Automate UI, please log out and visit https://a2-dev.test/
If all went well, you should see a block of disclosure text below the login, as well as a blue banner across the top of the screen.

You can then toggle the text of these, or show/hide by opening up the patch.toml file and changing the `show` attributes to false or changing the text.  In order to pick up the new changes you'll need to run `chef-automate config patch dev/patch.toml` again in hab studio.

### :white_check_mark: Checklist

**All PRs** must tick these:

- [x] I have read the [CONTRIBUTING document](https://github.com/chef/automate/blob/master/CONTRIBUTING.md).
- [x] All commits signed-off for the [Developer Certification of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

With occasional exceptions, all PRs **from Progress employees** must tick these:

- [x] Is the code clear? *(complicated code or lots of comments--subdivide and use well-named methods, meaningful variable names, etc.)*
- [x] Consistency checked? *(user notifications, user prompts, visual patterns, code patterns, variable names)*
- [x] Repeated code blocks eliminated? *(adapt and reuse existing components, blocks, functions, etc.)*
- [x] Spelling, grammar, typos checked? *(at a minimum use `make spell` in any component directory)*
- [x] Code well-formatted? *(indents, line breaks, etc. improve rather than hinder readability)*

All PRs **from Progress employees** should tick these if appropriate:

- [x] Tests added/updated? (all new code needs new tests)
- [ ] Docs added/updated? (all customer-facing changes)

*Please add a note next to any checkbox above if you are NOT ticking it.*

### :camera: Screenshots, if applicable

Banner and Disclosure text in Login
Screen Shot 2021-05-25 at 2 15 43 PM

Banner in Automate
Screen Shot 2021-05-25 at 2 16 07 PM

### CODE REVIEW

File: components/automate-dex/web/templates/login.html [hunk 1]
```diff
@@ -2,36 +2,67 @@
...
+  <meta charset="utf-8">
+  <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
+  <title>{{ issuer }}</title>
+  <meta name="viewport" content="width=device-width, initial-scale=1.0">
+  <link href='{{ url .ReqPath "static/main.css" }}' rel="stylesheet">
+  <link href='{{ url .ReqPath "theme/styles.css" }}' rel="stylesheet">
+</head>
+
+<body>
+  {{ if eq ("showBanner" | extra) "true" }}
```
Comment (SEAjamieD) [inline components/automate-dex/web/templates/login.html:16]: In these blocks, we're checking that the user has set their instance to show the banner, if it does, then we go ahead and display the banner.  You'll see this similar type of block used both in this file and in `password.html`.  Checking against `extra` is a pattern that comes from DexIDP (https://dexidp.io/docs/templates/)

File: components/automate-dex/web/templates/login.html [hunk 2]
```diff
@@ -2,36 +2,67 @@
...
+    <p class="warning-text" style='color: #{{ "bannerTextColor" | extra }}'>
+      {{ "bannerMessage" | extra }}
+    </p>
+  </div>
+  {{ end }}
+  <div class="theme-panel" role="main">
+    <div class="signin-login-container">
       <div class="logo">
-        <img class="theme-navbar__logo" alt="Chef Automate" src="{{url .ReqPath "static/img/automate-blue-d9789f4b.svg" }}">
+        <img class="theme-navbar__logo" alt="Chef Automate"
```
Comment (SEAjamieD) [inline components/automate-dex/web/templates/login.html:26]: You'll see a few changes like this as well - there is no change in the outcome.  Semgrep was complaining we were not interpolating the template variables properly, so we've updated while we're here.

File: components/automate-dex/web/templates/login.html [hunk 3]
```diff
@@ -2,36 +2,67 @@
...
+    const populateDisclaimer = () => {
+      const show = '{{ "showDisclosure" | extra }}';
+      if (show !== 'true') return;
+
+      const d = '{{ "disclosureMessage" | extra }}';
+      document.getElementById('disclosure').innerHTML = d;
+    }
+
+    populateDisclaimer();
+  </script>
```
Comment (SEAjamieD) [inline components/automate-dex/web/templates/login.html:65]: In order to get true HTML populating the disclosure block, we had to use JavaScript within the same file.  In this block, We are getting the disclosure text, and then appending it to the disclosure block.

File: components/automate-ui/src/app/app.module.ts [hunk 4]
```diff
@@ -354,7 +358,18 @@ import { WelcomeModalComponent } from './page-components/welcome-modal/welcome-m
...
+    {
+      provide: APP_INITIALIZER,
+      multi: true,
+      deps: [AppConfigService],
+      useFactory: (appConfigService: AppConfigService) => {
+        return () => {
+          return appConfigService.loadAppConfig();
+        };
+      }
+    }
```
Comment (SEAjamieD) [inline components/automate-ui/src/app/app.module.ts:372]: What this block does is allow us to call the `AppConfigService`'s `loadConfigApp` before the application renders.  In this case, we are using that time to grab our configuration for the warning banner.

File: components/automate-ui/src/app/services/app-config/app-config.service.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+})
+
+
+export class AppConfigService {
+
+  public appConfig: ConfigTypes;
+
+  constructor(private handler: HttpBackend) { }
+
+  public loadAppConfig() {
```
Comment (SEAjamieD) [inline components/automate-ui/src/app/services/app-config/app-config.service.ts:22]: This is the function that is doing all the heavy lifting.  It is reaching out to `/banner.js` and returning our configuration data

File: components/automate-ui/src/app/services/app-config/app-config.service.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    return new HttpClient(this.handler).get('/banner.js')
+      .toPromise()
+      .then(data => this.appConfig = data);
+  }
```
Comment (SEAjamieD) [inline components/automate-ui/src/app/services/app-config/app-config.service.ts:26]: This is the function that is doing all the heavy lifting.  It is reaching out to `/banner.js` and returning our configuration data

File: dev/banner_message.txt [hunk 7]
```diff
@@ -0,0 +1,10 @@
...
+
+<p>Nunc nibh eros, malesuada posuere molestie nec, pharetra eleifend magna. Nulla mollis vitae dui ac placerat. In nec pulvinar ex. Cras sodales at sapien finibus fermentum. Sed laoreet tempus velit sit amet condimentum. Ut sed risus id justo pulvinar sodales. Nunc eros velit, sollicitudin sit amet posuere vitae, vulputate in nunc. Maecenas quis risus maximus, finibus augue in, faucibus est.<p>
+
+<p>Nam aliquet maximus hendrerit. Duis lacinia faucibus vulputate. Donec id turpis arcu. Nullam pellentesque lobortis justo id lobortis. Orci varius natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Proin in ornare urna. Mauris id arcu enim. In nisi lectus, porta a aliquam in, porta sit amet urna. Suspendisse sollicitudin purus vitae lorem faucibus, a accumsan leo malesuada. Pellentesque egestas vitae nisi eu pulvinar. Proin magna arcu, ullamcorper eget venenatis ut, vehicula ac mi.</p>
+
+<p>Here is a link to <a href="https://google.com" rel="noopener noreferrer" target="_blank">google</a></p>
```
Comment (SEAjamieD) [inline dev/banner_message.txt:10]: This is our demo file for html.  Though it is a text file, a user will drop in their desired html for the disclosure panel.

File: dev/patch.toml [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+    show = true
+    message_file_path = "/src/dev/banner_message.txt"
+  [global.v1.banner]
+    show = true
+    message = "Changing the text around"
+    background_color = "3864f2" # Must be a hex value minus the hash symbol
+    text_color = "FFF" # Must be a hex value minus the hash symbol
+
+    # Find valid HEX codes at https://htmlcolorcodes.com/
+    # check your HTML validation at https://validator.w3.org/
```
Comment (SEAjamieD) [inline dev/patch.toml:12]: This is where a user will define their global configurations.

File: components/automate-ui/src/assets/app-config.json [hunk 9]
```diff
@@ -0,0 +1,6 @@
+{
+  "showBanner": true,
+  "bannerMessage": "This is my banner message that I want everyone to see",
+  "bannerBackgroundColor": "#7cc39f",
+  "bannerTextColor": "black"
+}
```
Comment (vinay033) [inline components/automate-ui/src/assets/app-config.json:6]: add EOL

**Review by vinay033 — APPROVED** [review]
LGTM 🎉

**Review by chaitali-mane — APPROVED** [review]
Changes looks good to me

**Pull request merged**